src/projection.jl

```diff
@@ -31,114 +37,352 @@ function Base.show(io::IO, project::ProjectTo{T}) where {T}
+    @eval ProjectTo(::$T) = ProjectTo{$T}()
+end
+ProjectTo(x::Integer) = ProjectTo(float(x))
+ProjectTo(x::Complex{<:Integer}) = ProjectTo(float(x))
 (::ProjectTo{T})(dx::Number) where {T<:Number} = convert(T, dx)
 (::ProjectTo{T})(dx::Number) where {T<:Real} = convert(T, real(dx))
 
-# Arrays 
-ProjectTo(xs::T) where {T<:Array} = ProjectTo{T}(; elements=map(ProjectTo, xs))
-function (project::ProjectTo{T})(dx::Array) where {T<:Array}
-    _call(f, x) = f(x)
-    return T(map(_call, project.elements, dx))
+# Arrays
+# If we don't have a more specialized `ProjectTo` rule, we just assume that there is
+# no structure worth re-imposing. Then any array is acceptable as a gradient.
+
+# For arrays of numbers, just store one projector:
+function ProjectTo(x::AbstractArray{T}) where {T<:Number}
+    element = ProjectTo(zero(T))
+    if element isa ProjectTo{<:AbstractZero}
+        return ProjectTo{NoTangent}() # short-circuit if all elements project to zero
+    else
+        return ProjectTo{AbstractArray}(; element=element, axes=axes(x))
+    end
 end
-function (project::ProjectTo{T})(dx::AbstractZero) where {T<:Array}
-    return T(map(proj -> proj(dx), project.elements))
+
+# In other cases, store a projector per element:
+function ProjectTo(xs::AbstractArray)
+    elements = map(_always_projector, xs)
+    if elements isa AbstractArray{<:ProjectTo{<:AbstractZero}}
+        return ProjectTo{NoTangent}() # short-circuit if all elements project to zero
+    elseif elements isa AbstractArray{<:ProjectTo{Any}}
+        return ProjectTo{AbstractArray}(; element=ProjectTo(), axes=axes(xs)) # ... or none project
+    else
+        # Arrays of arrays come here, and will apply projectors individually:
+        return ProjectTo{AbstractArray}(; elements=elements, axes=axes(xs))
+    end
 end
-(project::ProjectTo{<:Array})(dx::AbstractArray) = project(collect(dx))
 
-# Arrays{<:Number}: optimized case so we don't need a projector per element
-function ProjectTo(x::T) where {E<:Number,T<:Array{E}}
-    return ProjectTo{T}(; element=ProjectTo(zero(E)), size=size(x))
+function (project::ProjectTo{AbstractArray})(dx::AbstractArray{S,M}) where {S,M}
+    # First deal with shape. The rule is that we reshape to add or remove trivial dimensions
+    # like dx = ones(4,1), where x = ones(4), but throw an error on dx = ones(1,4) etc.
+    dy = if axes(dx) == project.axes
+        dx
+    else
+        for d in 1:max(M, length(project.axes))
+            size(dx, d) == length(get(project.axes, d, 1)) || throw(_projection_mismatch(project.axes, size(dx)))
+        end
+        reshape(dx, project.axes)
+    end
+    # Then deal with the elements. One projector if AbstractArray{<:Number},
+    # or one per element for arrays of arrays:
+    dz = if hasfield(typeof(backing(project)), :element)
+        T = project_type(project.element)
+        S <: T ? dy : map(project.element, dy)
+    else
+        map((f,y) -> f(y), project.elements, dy)
+    end
+    return dz
```

Comment:
I unified the handling of arrays of arrays, and arrays. The run the same reshape. `project.element` means one, `project.elements` means one per element. 

`Any[1,2,3]` is treated like an array of arrays. `Any[:a, :b, :c]` will start the same, take a shortcut that gives `project.element == ProjectTo{Any}()` (just one), and since `S <: Any` this will do nothing.

Reply:
If we had a FillArrays dependency could we have just one path?
Because `Array{T} where T<:Number}` could set the `elements=FillArray(ProjectTo(zero(T)))` ?

Reply:
The really happy path here is `S <: T ? dy`, which doesn't copy the array. One would still need to allow that, somehow. And seems a bit cleaner to only store one if you only need one, even if really Fill is ultimately one.